cocotb/share/lib/gpi_log/gpi_logging.c

```diff
@@ -145,54 +155,65 @@ void gpi_log(const char *name, long level, const char *pathname, const char *fun
     if (level < local_level)
         return;
 
-    // Ignore truncation
-    // calling args is level, filename, lineno, msg, function
-    //
     PyGILState_STATE gstate = PyGILState_Ensure();
 
-    PyObject *check_args = PyTuple_New(1);
-    PyTuple_SetItem(check_args, 0, PyLong_FromLong(level));
+    PyObject *filename_arg = NULL, *lineno_arg = NULL, *msg_arg = NULL, *function_arg = NULL;
+
+    PyObject *level_arg = PyLong_FromLong(level);                  // New reference
+    if (level_arg == NULL)
+        goto error;
+
+    PyObject *filter_ret = PyObject_CallFunctionObjArgs(pLogFilter, level_arg, NULL);
+    if (filter_ret == NULL)
+        goto error;
 
-    PyObject *filter_ret = PyObject_CallObject(pLogFilter, check_args);
-    Py_DECREF(check_args);
-    if (filter_ret == NULL) {
-        PyErr_Print();
-        PyGILState_Release(gstate);
-        return;
-    }
     int is_enabled = PyObject_IsTrue(filter_ret);
     Py_DECREF(filter_ret);
     if (is_enabled < 0) {
         /* A python exception occured while converting `filter_ret` to bool */
-        PyErr_Print();
-        PyGILState_Release(gstate);
-        return;
+        goto error;
     }
 
     if (!is_enabled) {
-        PyGILState_Release(gstate);
-        return;
+        goto ok;
     }
 
+    // Ignore truncation
     va_start(ap, msg);
     n = vsnprintf(log_buff, LOG_SIZE, msg, ap);
     va_end(ap);
 
-    PyObject *call_args = PyTuple_New(5);
-    PyTuple_SetItem(call_args, 0, PyLong_FromLong(level));           // Note: This function steals a reference.
-    PyTuple_SetItem(call_args, 1, PyUnicode_FromString(pathname));   // Note: This function steals a reference.
-    PyTuple_SetItem(call_args, 2, PyLong_FromLong(lineno));          // Note: This function steals a reference.
-    PyTuple_SetItem(call_args, 3, PyUnicode_FromString(log_buff));   // Note: This function steals a reference.
-    PyTuple_SetItem(call_args, 4, PyUnicode_FromString(funcname));
-
-    PyObject *handler_ret = PyObject_CallObject(pLogHandler, call_args);
-    Py_DECREF(call_args);
-    if (handler_ret == NULL){
-        PyErr_Print();
-        PyGILState_Release(gstate);
-        return;
-    }
+    filename_arg = PyUnicode_FromString(pathname);      // New reference
+    if (filename_arg == NULL)
+        goto error;
```

Comment:
nit: up to @imphil. but braces are probably a good idea here.

Reply:
Added braces to the if statements.